### ISSUE TITLE
Migrate build to pyproject.toml

### ISSUE THREAD

**Issue by Technologicat** [Technologicat/mcpyrate#42]
The migration seems more complex than necessary. Waiting until someone writes an ELI5 migration guide.

An added point of complexity here is that we need to build the wheels either a) without compiling bytecode, or b) with *macro-enabled* bytecode, which requires a custom build command. And running custom commands in the installer is frowned upon (mostly for good reason, but here such a feature would be useful).

Seems we're not the only one in this boat, though.

*You are telling me that we had a working `python setup.py` solution for 10+ years, this workflow is now quasi deprecated, and the recommended replacement is **it's complicated**?! I'm just trying to get my package modernized. Why does that need to be **complicated**?*
  --[Gregory Szorc](https://gregoryszorc.com/blog/2023/10/30/my-user-experience-porting-off-setup.py/)

**Comment by alextremblay** [comment]
looking at your setup.py file, it looks like your requirements for a pyproject.toml file are not too complex (especially compared to some other projects I've seen)

the only form of dynamism i see in your project is the process of extracting the version from the `__version__` field of `__init__.py`

Given that, i think you have a lot of options.

Rather than an ELI5 migration guide, I think perhaps it would be easier to show you. 

I will submit a draft PR outlining a recommended approach shortly

**Comment by Technologicat** [comment]
Thanks for your interest, and the PR!

At a first glance, looks great. I'll review this in more detail as soon as I can, likely on Monday.

One specific question: is it possible to disable `.pyc` precompilation in the `pyproject/pdm` backend, or to force to use a custom compile command?

This is something that was a problem with `setuptools`/`pip` - the `mcpyrate` modules would automatically get precompiled without macro support. Then, when that packaged `mcpyrate` was installed, and another project (e.g. the sister project [`unpythonic`](https://github.com/Technologicat/unpythonic)) attempted to import the quasiquote macros from `mcpyrate`, it would error out.

**Comment by alextremblay** [comment]
Already taken care of. 
The `excludes = ["**/test", "**/__pycache__"]` line in `pyproject.toml` prevents any .pyc files from being included in the sdist and wheel files

You can confirm this is the case by looking at the breakdown of file contents in the PR description and see that no `.pyc` files or `__pycache__/*` files are included in either 

:)

**Comment by Technologicat** [comment]
That's great, then!

I'll review the PR today.

**Comment by Technologicat** [comment]
Hmm. To think of it, that's not sufficient. My previous sdist and wheel files don't have any `.pyc` files, either.

Rather, the issue is rather that `pip` generates these `.pyc` files (in an incorrect way) when invoking `pip install mcpyrate`. One must instead use `pip install mcpyrate --no-compile` to prevent the incorrect `.pyc` files from being generated and installed at install time.

To precompile macro-enabled code correctly, `mcpyrate` must be loaded. But the standard `py3compile` tool used by `pip` doesn't do that. (`py3compile` is essentially a wrapper for the `py_compile` stdlib module. Or at least used to be when I last looked at it.)

Is there anything in pyproject/pdm that can help here?

Either disable install-time precompilation at the package level, or customize how install-time precompilation is done for this particular package?

This issue is unique to macro expanders and macro-enabled packages, as anything else can be precompiled in the standard way.

**Comment by alextremblay** [comment]
Ahhh, interesting...

How does mcpyrate handle this now? looking at `setup.py`, `setup.cfg`, and `makedist.sh`, i don't see any mechanism to influence how pip installs the resultant package, and i'm not aware of the existence of any package metadata we can add to influence how pip installs said package...

pdm has a number of build hooks we can hook into to influence things, and i know at least some of them get bundled into the sdist and executed during sdist installation, but for wheel files things get trickier, i don't know if there is any mechanism to trigger code to run during the installation / unpacking of a wheel file (AFAIK that's kind of the point of wheel files haha)

**Comment by alextremblay** [comment]
I took a look at the source for https://github.com/pypa/pip

followed the rabbit-hole of the no-compile flag from [src/pip/_internal/commands/install.py#L222](https://github.com/pypa/pip/blob/f2cdcd04e11fbe27ea161f3b3868f6a27fbb0f53/src/pip/_internal/commands/install.py#L222) all the way through to where it gets used: [src/pip/_internal/operations/install/wheel.py#L609](https://github.com/pypa/pip/blob/f2cdcd04e11fbe27ea161f3b3868f6a27fbb0f53/src/pip/_internal/operations/install/wheel.py#L609)

I see no place in the code path where this value can get influenced by the contents of the package being installed...

interesting note: the `--compile`/`--no-compile` flag is only used in this function, but don't let its name fool you. `_install_wheel` is called to install both wheel and sdist files, and also called when doing modern editable installs. The only time this function isn't called is when doing what pip calls "legacy editable installs", compatibility for which is set to be removed in the next pip release

**Comment by Technologicat** [comment]
Merged your PR, which resolves this, so closing. Thanks again for your contribution!

**Comment by Technologicat** [comment]
One additional note: the CI scripts in `.github/workflows` required some work to get them working correctly with PDM.

They have now been fixed too (for both `mcpyrate` as well as for `unpythonic`).
